zend_abstract_interface/config/php5/config_ini.c

```diff
@@ -0,0 +1,334 @@
+
+    TSRMLS_FETCH();
+
+    for (zai_config_id i = 0; i < zai_config_memoized_entries_count; i++) {
+        zai_config_memoized_entry *memoized = &zai_config_memoized_entries[i];
+        for (uint8_t n = 0; n < memoized->names_count; n++) {
+            zai_config_name *ini_name = &ini_names[i * ZAI_CONFIG_NAMES_COUNT_MAX + n];
+            zai_string_view name = {.len = memoized->names[n].len, .ptr = memoized->names[n].ptr};
+            zai_config_add_ini_entry(memoized, name, ini_name, module_number, i TSRMLS_CC);
+            int ini_available =
+                zend_hash_find(EG(ini_directives), ini_name->ptr, ini_name->len, (void **)&memoized->ini_entries[n]);
+            UNUSED(ini_available);  // compiled away asserts cause compile errors otherwise...
+            assert(ini_available == SUCCESS);
+        }
+    }
+
+#if ZTS
+    original_thread_end_handler = tsrm_set_new_thread_end_handler(zai_config_lock_ini_copying);
+#endif
+}
+
+#if ZTS
+void zai_config_ini_rinit() {
+    if (!env_to_ini_name) return;
+
+    TSRMLS_FETCH();
+
+    // we have to cover two cases here:
+    // a) update ini tables to take changes during first-time rinit into account
+    // b) apply and verify user.ini/htaccess settings
+    for (uint8_t i = 0; i < zai_config_memoized_entries_count; ++i) {
+        zai_config_memoized_entry *memoized = &zai_config_memoized_entries[i];
+        bool applied_update = false;
+        for (uint8_t n = 0; n < memoized->names_count; ++n) {
+            zend_ini_entry *source = memoized->ini_entries[n], *ini;
+            zend_hash_find(EG(ini_directives), source->name, source->name_length, (void **)&ini);
+            if (ini->modified) {
+                if (ini->orig_value == ini->value) {
+                    ini->value = source->value;
+                }
+                ini->orig_value = source->value;
+                ini->orig_value_length = source->value_length;
+
+                if (!applied_update) {
+                    if (ZaiConfigOnUpdateIni(ini, ini->value, ini->value_length, NULL, NULL, NULL,
+                                             PHP_INI_STAGE_RUNTIME TSRMLS_CC) == SUCCESS) {
+                        // first encountered name has highest priority
+                        applied_update = true;
+                    } else {
+                        efree(ini->value);
+                        ini->value = ini->orig_value;
+                        ini->value_length = ini->orig_value_length;
+                        ini->modified = false;
+                        ini->orig_value = NULL;
+                    }
+                }
+            } else {
+                ini->value = source->value;
+                ini->value_length = source->value_length;
+            }
+        }
+    }
+}
+#endif
+
+void zai_config_ini_mshutdown() {
+    used_original_ini_values = 0;
+    for (int i = 0; i < used_original_ini_values; ++i) {
+        free(original_ini_values[i]);
+    }
```

Comment:
you mean none? :-D Yep, fixing. (I'm sure that was correct at some point and then I somehow broke it)

Reply:
Haha! Yes, I mean none. 🤦

I'm curious why none of the leak detection tests didn't catch this?

Reply:
It's malloc() memory, not emalloc

Reply:
~Sure, but ASan should have detected this, no?~ Ah! We don't have an ASan build on PHP 5. 😞 We definitely need to add valgrind support/some other leak detection to the ZAI tests for non-ASan builds (in a separate PR of course). And I'm assuming this didn't show up in the existing phpt leak-detection tests because this code path is never hit in any of those tests?

Reply:
And the latter is correct: because there is no code using ZAI config for PHP 5 yet. Only our zai testsuite.